Bio/SeqRecord.py

```diff
@@ -238,10 +239,10 @@ def __init__(
         # annotations about parts of the sequence
         if features is None:
             features = []
-        elif not isinstance(features, list):
-            raise TypeError(
-                "features argument should be a list (of SeqFeature objects)"
-            )
+        elif not isinstance(features, list) and all(
```

Comment:
Does it have to be a list? For example, should a tuple or iterator be excluded? You could call `list(features)` and raise a `TypeError` only if that operation fails. That would be more in line with Python's duck-typing principle.

Reply:
Historically we've only put a list there, but conversion of the passed argument to a list with ``self.features = list(features)`` might be OK.

(Long term any attempt at a lazy-loading ``SeqRecord`` would likely have a proxy object here behaving like a list)

Reply:
@peterjc we can't type check the returns of a generator in this way without loosing the lazy-loading property. The method proposed by @mdehoon would result in the conversion of the generator into a list. 

So, we can test for the correct data structure, we can convert it and see what happens or we can not test at all. Or some combination of these three variants depending on the features type. e.g. If it is a tuple, convert to list, then check the content if it is a list (or was a tuple), skip the testing if it is a generator (to keep the lazy loading) or consume the generator and then proceed as with the list.

I would probably go with the list requirement, this way, when a generator or proxy object is programmed at a later point in time, this check will immediately raise a red flag and the type checking can be adapted to the new features.

Reply:
I was not trying to say that the list conversion blocks the development of lazy loading of SeqFeates, but that the conversion into a list may remove the benefits of lazy loading. Future developers of a lazy loading SeqFeature object might not be aware of the intricacies of this type check and thus fail to see the necessity to adapt SeqRecord in a way that it doesn't load all the SeqFeatures just for type checking. I just think the list conversion increases the chance of unexpected behavior.

Also, the requirement that features be a list is baked into the code since [13 years](https://github.com/biopython/biopython/blame/64de71882dc465a6a9e1ab600987109883f40bad/Bio/SeqRecord.py). I don't think it is desirable to remove this check now without clear need and motivation. It's an API change after all and should only happen with good reason. Some people might rely on this check. And a hypothetical future changes in SeqFeature is no good reason to introduce that conversion now either - at least in my thinking.

My conclusion is that the list doesn't solve more problems than it creates and thus my vote is on the status quo. 

However, if you insist on this conversion, it should be worth a separate PR and possibly part of a general attempt to make biopython more pythonic. This PR is about the type checking the contents of the list which is the features variable. The type conversion of the features variable itself is a separate issue.

Reply:
If we weren't checking for `SeqFeature` before and with this PR we are requiring that they are all `SeqFeature`s, isn't that in itself an API change?